Server/PlayerLogic.py

```diff
@@ -2,58 +2,151 @@
+    def update(self):
+        messages = Server.get_instance().get_targets_messages(climess.Target.PLAYER_LOGIC+str(self._id))
+        self._process_requests(messages)
+        self._do_physics()
+        self._send_updated_pos()
 
-    def _process_requests(self, network_messages):
-        for m in network_messages:
+    def get_id(self):
+        return self._id
+
+    def _process_movement_messages(self, pos_mess):
+        for m in pos_mess:
             for mess in m.movement_list:
                 if mess == climess.ActionRequest.MOVE_LEFT:
                     self._move_left()
                 elif mess == climess.ActionRequest.MOVE_RIGHT:
                     self._move_right()
 
-    def _move_left(self):
-        self._x = self._x - self._speed
-
-        self._x = min(max(self._x, self.X_MIN), self.X_MAX)
-        self._y = min(max(self._y, self.Y_MIN), self.Y_MAX)
-
-    def _move_right(self):
-        self._x = self._x + self._speed
-
-        self._x = min(max(self._x, self.X_MIN), self.X_MAX)
-        self._y = min(max(self._y, self.Y_MIN), self.Y_MAX)
+    def _process_requests(self, network_messages):
+        position_messages = list(filter(lambda x: x.type == climess.MessageType.PLAYER_MOVEMENT, network_messages))
+        self._process_movement_messages(position_messages)
 
     def _send_updated_pos(self):
         msg = BaseMessage(mess_type=sermess.MessageType.PLAYER_POSITION, target=sermess.Target.PLAYER + str(self._id))
         msg.player_id = self._id
-        msg.x = self._x
-        msg.y = self._y
+        msg.x = self._pos.x
+        msg.y = self._pos.y
         Server.get_instance().send_all(msg)
 
-    def update(self):
-        messages = Server.get_instance().get_targets_messages(climess.Target.PLAYER_LOGIC+str(self._id))
-        position_messages = list(filter(lambda x: x.type == climess.MessageType.PLAYER_MOVEMENT, messages))
-        if position_messages:
-            self._process_requests(position_messages)
-            self._send_updated_pos()
-        if self.initial_pos_sending:
-            self._send_updated_pos()
-            self.initial_pos_sending = False
+    def _add_force(self, force2d):
+        self._forces.append(force2d)
 
-    def get_id(self):
-        return self._id
+    def _add_ground_directed_force(self, force_mag, direction):
+        ang = self._terrain.get_angle_rad(self._pos.x)
+        force = Vector2D.mag_ang_init(force_mag*direction.value, ang)
+        self._add_force(force)
+
+    def my_dir(self):
+        if self._vel.x > 0:
+            return Direction.RIGHT
+        return Direction.LEFT
+
+    def can_accelerate(self, direction=1):
+        if direction*self._vel.x < 0:  # it can always decelerate
+            return True
+        base_max_vel = 3.0
+        angle_dependency = 3.0
+        angle_bonus = -direction * math.sin(self._terrain.get_angle_rad(self._pos.x)) * angle_dependency
+        return self._vel.mag() < base_max_vel + angle_bonus
```

Comment:
I think it would be better if _base_max_vel_ and _angle_dependency_ would be constants.

Reply:
True